File: python/src/threads/sandbox/tree/source.py
"""Exact reads over a chunked async byte source."""

import inspect
from collections.abc import AsyncIterable, Callable


class Source:
    """Pulls exact byte counts from chunks of any size; `offset` counts the bytes taken."""

    def __init__(self, source: AsyncIterable[bytes]) -> None:
        self._it = aiter(source)
        # A view, so taking a piece never copies the rest of the chunk.
        self._head = memoryview(b"")
        self.offset = 0

    async def next(self, most: int) -> bytes | None:
        """The next bytes, at most `most`; None at the end of the source."""
        while not self._head:
            try:
                self._head = memoryview(await anext(self._it))
            except StopAsyncIteration:
                return None
        piece, self._head = self._head[:most], self._head[most:]
        self.offset += len(piece)
        return piece.tobytes()

    async def take(self, n: int, each: Callable[[bytes], None]) -> bool:
        """Passes the next `n` bytes to `each`; False when the source ends first."""
        left = n
        while left > 0:
            piece = await self.next(left)
            if piece is None:
                return False
            each(piece)
            left -= len(piece)
        return True

    async def exact(self, n: int) -> bytes | None:
        """The next `n` bytes; None when the source ends first."""
        out = bytearray()
        return bytes(out) if await self.take(n, out.extend) else None

    async def close(self) -> None:
        """Stops a generator source early, so its producer can release what it holds."""
        it: object = self._it
        if inspect.isasyncgen(it):
            await it.aclose()

File: typescript/packages/core/src/sandbox/tree/source.ts
/** Exact reads over a chunked byte source; `offset` counts the bytes taken. */
export class Source {
  offset = 0;
  #head: Uint8Array = new Uint8Array();
  readonly #it: AsyncIterator<Uint8Array>;

  constructor(source: AsyncIterable<Uint8Array>) {
    this.#it = source[Symbol.asyncIterator]();
  }

  /** The next bytes, at most `max`; undefined at the end of the source. */
  async next(max: number): Promise<Uint8Array | undefined> {
    while (this.#head.length === 0) {
      const next = await this.#it.next();
      if (next.done === true) return undefined;
      this.#head = next.value;
    }
    const piece = this.#head.subarray(0, max);
    this.#head = this.#head.subarray(piece.length);
    this.offset += piece.length;
    return piece;
  }

  /** Passes the next `n` bytes to `each`; false when the source ends first. */
  async take(n: number, each: (piece: Uint8Array) => void): Promise<boolean> {
    for (let left = n; left > 0; ) {
      const piece = await this.next(left);
      if (piece === undefined) return false;
      each(piece);
      left -= piece.length;
    }
    return true;
  }

  /** The next `n` bytes; undefined when the source ends first. */
  async exact(n: number): Promise<Uint8Array | undefined> {
    const out = new Uint8Array(n);
    let at = 0;
    const whole = await this.take(n, (piece) => {
      out.set(piece, at);
      at += piece.length;
    });
    return whole ? out : undefined;
  }

  /** Stops the source early, so its producer can release what it holds. */
  async close(): Promise<void> {
    await this.#it.return?.();
  }
}
